### ISSUE TITLE
Add ability to rename a contract file

### ISSUE THREAD

**Issue by filippsen** [SuperblocksHQ/superblocks-lab#99]
## Initial issue
> There is an issue [...] when renaming a contract, even when all tabs except _Configure_ are closed and even after reloading the application.

> [...] changing the name from `a` to `b` and then renaming it back would cause a problem.

### Result
`Error: Could not rename contract source file. Please close the tab containing the contract’s source code and try again.`

### Suspects
> [...] might be related to [out of sync local storage], as the name does not change in the `project.files.contracts`.

> Files and dappfile are out of sync in local storage.

### First try on reproducing the problem

Unable to find it or reproduce at first. The renaming actions always reflect to _dapps1.0.0._ state (inspected via browser debugger). The only times it returns an error is when the file is open. The error message asks to close the tab and retry; then it works as intended. Some samples:

*Requirements:*
Create a _New project_ from _Hello World_ template.

*Check initial _Application Local Storage_ (dapps1.0.0) state:*
>contracts:{type: "d", children:{HelloWorld.sol: {type: "f",…}} [...]

* Rename `HelloWorld` to `b`:
>contracts:{type: "d", children:{b.sol: {type: "f",…}} [...]

* Rename `b` to `HelloWorld`:
>contracts:{type: "d", children:{HelloWorld.sol: {type: "f",…}} [...]

* Rename `HelloWorld` to `a`:
>contracts:{type: "d", children: {a.sol: {type: "f",…}}} [...]

* Rename `a` to `b`:
>contracts:{type: "d", children: {b.sol: {type: "f",…}}} [...]

* Rename `b` to `a`:
>contracts:{type: "d", children: {a.sol: {type: "f",…}}} [...]

## Second pass
> problem when opening the _Deploy_ tab, then the _Configure_ tab for renaming. It breaks on _Save_, sometimes ...

### Second try on reproducing the problem
1. Create a _New project_ from _Hello World_ template.
2. Now, proceed with the following actions, always accessing it from the left side menu:
3. Expand `HelloWorld.sol` in the _Files_ hierarchy;
4. Click *Configure*: rename `HelloWorld` to `b`. Save;
5. Expand `b.sol`;
6. Click *Configure*. Do nothing. Leave the tab open.
7. Click *Compile*. Leave the tab open. (Could also be triggered from the editor menu)
8. Click *Deploy*. Leave the tab open. (Could also be triggered from the editor menu)
9. Click *Configure*. Rename back to `HelloWorld`

### Resulted behavior
1. _Application local storage_ gets updated as expected, including the file data: `contracts:{type: "d", children: {HelloWorld.sol: {type: "f",…}}}`;
2. The rendered contract file name doesn't get updated: `b.sol`;
3. An error message pops up: `Could not load contract source code. Is there any contract not saved?`
4. `b.sol` contents are "lost" or overwritten as if it was a "New File" with empty "contract b {}" data. Actually, we know the file was renamed to HelloWorld.sol and the contents are still there (see the first point - 1).

### Explaning the results
It looks at first that the exact problem happened only when having the *Configure* window open while running a *Compile* or *Deploy* action (or both) which results in an error (red output message). It can be further compressed by repeating steps 5-9 directly from the starting _Hello World_ template.
When the rename is triggered, upon hitting the *Save* button in *Configure*, the data is apparently updated but the on-screen reference isn't (as Javier had pointed out: it is out of sync).

### Other information
In `projecteditor/editor-contract`, the `save` function gets called as expected, and it ends with `finalize` and `status==0`. The answer was expected to be around that area, considering it gets triggered by the Edit/Save button. Failed to spot divergences in the code itself.

## Third pass
> it seems like it only happens when windows associated with the contract are open

> windows should not be open before renaming

### Another idea ...
> consider automatically closing all windows before saving

### ... another stone ...
> it would be required to re-open Config *after* closing the windows, because the configuration could be "tainted"

### ... another draft
A less intrusive patch and apparently more "correct" approach, considering the nature of the bug, would be to always close the _Configure_ tab whenever opening a new window. It would be more "correct", considering it has to do with sequencing and combinations, which we probably haven't mapped all out extensively. That is, if it would suffice by itself. Let's analyze if that is indeed the case . . .

The first question then was: how would you pick just the _Configure_ pane from inside the `_openItem` scope ?

Start with code in `src/components/projecteditor/control/control.js`. Consider the following addition to `_openItem`, before calling `this.props.router.panes.openItem(item)`, in line 505:

`this.props.router.panes.closeAll();`

This will close *everything* before opening a new window, right ? This is the "inverse" of the previous approach, which prevents the "bug" in *most* cases. It's been mentioned *most* cases because it would still be possible to trigger it, if one is fast enough, by opening up the _Configure_ window while the "lightning" icon is  spinning during _Deploy_, for instance.
With that said, change the proposed `closeAll` addition to a `close("Configure")` equivalent and it could be eligible as a first patch.

## Current state
*Commit:* https://github.com/SuperblocksHQ/superblocks-lab/commit/4e97a3408c61f1d0ccbaba845984cff14dbf83a2  
*Branch:* `prevent-rename-bug`

The code tries to expand on what the `close("Configure")` would look like. The current code apparently prevents the appearance of the rename issue.

Expected to work for some cases, according to the following tests:  
*Reference branch*: `prevent-rename-bug`
*Main theme*:
>test if the bug doesn't appear if the compile/deploy/interact windows are manually closed before renaming the contract

*Test 1* - _from New Hello World project_
*Variant*: everything is open from the left side menu and just the contract is manually closed
* *Open* _HelloWorld.sol_
* *Click* _Compile_ from the left side menu
* *Click* _Deploy_ from the left side menu
* *Click* _Interact_ from the left side menu
* *Click* _Configure_ -> try to rename to _b_ -> fails because _HelloWorld.sol_ is open
* *Close* _HelloWorld.sol_
* *Click* _Configure_ -> rename to _b_
Works

*Test 2* - _from New Hello World project_
*Variant:* everything is open from the left side menu and Close all shortcut is used
* *Open* _HelloWorld.sol_
* *Click* _Compile_ from the left side menu
* *Click* _Deploy_ from the left side menu
* *Click* _Interact_ from the left side menu
* *Click* any tab with the right mouse button and select _Close all_
* *Click* _Configure_ -> rename to _b_
Works

*Test 3* - _from New Hello World project_
*Variant:* everything is open from the left side menu and everything is manually closed
* *Open* _HelloWorld.sol_
* *Click* _Compile_ from the left side menu
* *Click* _Deploy_ from the left side menu
* *Click* _Interact_ from the left side menu
* *Close* all 4 open tabs manually
* *Click* _Configure_ -> rename to _b_
Works

*Test 4* - _from New Hello World project_
*Variant:* everything is open from the editor bar and the single tab is manually closed
* *Open* _HelloWorld.sol_
* *Click* _Compile_ from the editor bar
* *Click* _Deploy_ from the editor bar
* *Click* _Interact_ from the editor bar
* *Close* _HelloWorld.sol_ tab by clicking the X
* *Click* _Configure_ -> rename to _b_
Works

Reported to not be working by reviewer:
> the window closes, but the renaming doesn't seem to work

**Comment by filippsen** [comment]
## (Repeating) steps to reproduce
1. Create _New project_ from _Hello World_ template
2. Expand _HelloWorld.sol_ file
3. Open _Configure_
4. Open _Compile_
5. Open _Deploy_
6. Open _Configure_ again (switch focus to)
7. _Rename_ from `HelloWorld` to `HelloWorld2`
8. _Save_

## (Further) troubleshooting 

### When `HelloWorld.sol` window is open
```
editor-contract: project rename file - original source path:  /contracts/HelloWorld.sol  | file:  HelloWorld2.sol  | status:  1
```

### When closing `HelloWorld.sol` only (keeping remaining windows open)
```
editor-contract: project rename file - original source path:  /contracts/HelloWorld.sol  | file:  HelloWorld2.sol  | status:  0
```

### Commenting out the `finalize` function call in `editor-contract.js`
```
editor-contract: project rename file - original source path:  /contracts/HelloWorld.sol  | file:  HelloWorld2.sol  | status:  0
```

* Opening `HelloWorld.sol` shows as if the file was overwritten by the _New contract_ template.  
* Inspecting _Local Storage_: `HelloWorld2.sol:{type: "f",…}`
* `HelloWorld.sol` is no more because it was renamed by the successful `this.props.project.renameFile` call

### Inspecting the Files hierarchy
```
editor-contract.js:143 editor-contract: save -  
MouseEvent {isTrusted: true, persist: ƒ, nativeEvent: MouseEvent, screenX: 539, screenY: 455, …}
control.js:129 control: renameFile -  
{dir: "a", inode: 7546362, dappfile: {…}, _filecache: Array(0)}
 /contracts/HelloWorld.sol HelloWorld2.sol
backend.js:231 backend: renameFile -  a /contracts/HelloWorld.sol HelloWorld2.sol
backend.js:274 backend: renameFile - data  
{projects: Array(9)}
editor-contract.js:201 editor-contract: project rename file - original source path:  /contracts/HelloWorld.sol  | file:  HelloWorld2.sol  | status:  0
editor-contract.js:171 editor-contract: finalize
control.js:109 
[{…}]
0: 
args: [{…}] 
blockchain: "ethereum" 
name: "HelloWorld2"
network: "browser"
source: "/contracts/HelloWorld2.sol"
__proto__: Object
length: 1
__proto__: Array(0)
```


* In `control.js`, `_saveProject` receives the old `projectItem` when the deploy or compile windows are open i.e. `HelloWorld.sol` instead of `HelloWorld2.sol`.

### Inspecting `projectItem.props.state.data.dappfile.root.contracts`

#### On successful rename when no windows are open:
```
0: 
args: [{…}]
blockchain: "ethereum"
name: "HelloWorld2"
network: "browser"
source: "/contracts/HelloWorld2.sol"
__proto__: Object
length: 1
__proto__: Array(0)
```

#### When compile and deploy windows are open:
```
0:
args: [{…}]
blockchain: "ethereum"
name: "HelloWorld"
network: "browser"
source: "/contracts/HelloWorld.sol"
__proto__: Object
length: 1
__proto__: Array(0)
```

**Comment by filippsen** [comment]
Upon further inspection, it appears that`reloadProjects` touches storage data which affects _Configure_ window. The side-effects are unknown. Reference: `this.props.router.control._reloadProjects()`

Considering that, removing `reloadProjects` calls from `compiler.js` and `deployer.js` apparently remediates the problem.
The next question would then be: what are the consequences of not calling `reloadProjects` in the context of _Compile_ and _Deploy_ ?

**Comment by filippsen** [comment]
Proposed patch: https://github.com/Bashlund/superblocks-lab/commit/d20e97a68732a2b9d6ce24c132938b1de099800c

**Comment by filippsen** [comment]
Preliminary tests show the proposed patch does solve the renaming issue, but adds traction to the original renaming action flow, which used to be straightforward.

Noticeable changes:
* The contract items used to collapse after renaming
* The following message is now present: "Please close any editor, compile or deploy window which is open for this contract, then try again to rename it."

**Comment by filippsen** [comment]
New proposed patch: https://github.com/Bashlund/superblocks-lab/commit/61729f8eb9da649ea5a4f4df25a33360cc886d7d

**Comment by filippsen** [comment]
Solved with request #168
